dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,595 @@
+  }
+
+  bool IsExpanding() const {
+    return is_expanding_;
+  }
+
+  bool IsExpandable() const {
+    auto expand_like_idx = GetReferenceInputIdx();
+    assert(expand_like_idx == -1 || GetInputExpandDesc(expand_like_idx).NumDimsToExpand() > 0);
+    return expand_like_idx >= 0;
+  }
+
+ protected:
+  /**
+   * @brief Controls if the SequenceOperator should expand batches in the current iteration.
+   *
+   * It can be used to restrict cases when the expansion is performed, but it is an error for
+   * ``ShouldExpand`` to return true if none of the inputs is expandable (i.e. !IsExpandable()).
+   * Overriding the method may be useful for validation of the initial input layouts.
+   */
+  virtual bool ShouldExpand(const workspace_t<Backend> &ws) {
+    return IsExpandable();
+  }
+
+  /**
+   * @brief A common path for processing inferred output shapes for expanding and non-expanding
+   * cases.
+   *
+   * By default, the expanded output is assumed to have the same expandable layout and
+   * expandable shape as the expandable input. The inferred shapes are coalesced, i.e. the
+   * SequenceOperator verifies if the the shapes of frames corresponding to the same sequence are
+   * equal and merges them into sequences.
+   *
+   * May be overriden to infer output shapes outside of the operator's SetupImpl method,
+   * which may be easier or more performant.
+   */
+  virtual bool ProcessOutputDesc(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<Backend> &ws, bool is_inferred) {
+    if (is_inferred && IsExpanding()) {
+      CoalesceOutputShapes(output_desc, ws);
+    }
+    return is_inferred;
+  }
+
+  /**
+   * @brief Controls if given input should expand leading channel dimensions.
+   *
+   * If false, only a batch whose layout starts with ``F`` will be considered expandable
+   * (for example ``FHWC -> HWC``). If true, layouts starting with ``C`` will be expandable as well,
+   * i.e. ``FCHW -> HW, CFHW -> HW, CHW -> HW`` etc.
+   *
+   * However, operator does not support specifing tensor arguments per-chanel.
+   */
+  virtual bool ShouldExpandChannels(int input_idx) const {
+    (void)input_idx;
+    return false;
+  }
+
+  /**
+   * @brief Gets the ExpandDesc instance that describes the expandable prefix of the output layout
+   * and corresponding prefixes of the shapes in the batch. By default, the sequence shape and
+   * expandable layout are assumed to be the same for the output and the expandable input. If that's
+   * not the case, the method may be overriden to provide different ExpandDesc instance.
+   */
+  virtual const ExpandDesc &GetOutputExpandDesc(const workspace_t<Backend> &ws,
+                                                int output_idx) const {
+    (void)output_idx;
+    return GetInputExpandDesc(GetReferenceInputIdx());
+  }
+
+  virtual int GetArgExpandDescInputIdx(const std::string &arg_name) const {
+    (void)arg_name;
+    return GetReferenceInputIdx();
+  }
```

Comment:
Let's put the doc here also. To be consistent and all ;)

Reply:
done